Integrate graphql crunch

zephraph opened the pull request:
[Graphql Crunch](https://github.com/banterfm/graphql-crunch) is a library that flattens deeply nested object hierarchies into an array and deduplicates references in the array. This _could_ yield some nice performance benefits, especially when used with complex queries. 

This implementation follows their docs example and only crunches when the request has the `crunch` query param. It could just as easily be added as a header though. Ultimately this allows for incremental adoption across the Artsy clients. 

I'd like to add some sort of validation around this functionality and it'd also be beneficial to benchmark against some actual Artsy queries.

zephraph commented on src/index.js:
```diff
@@ -124,6 +125,12 @@ async function startApp() {
           isProduction,
         }),
+        formatResponse: resp => {
+          if (req.query.crunch && resp.data && !resp.data.__schema) {
```
Only apply when

1. `crunch` is supplied as a query param
2. There's actual data to return
3. The current response isn't an introspection query

mzikherman commented:
This looks cool! I like the 'opt-in' pattern, it can thus be safely deployed w/o any feature flag (although a spec for the behavior as you mention would be nice).

We certainly have lots of deeply nested queries so it would be interesting to see what the results of some benchmarks look like.

Congrats on the first PR!

zephraph commented:
I was going down the path of writing tests and ran into a few interesting issues. 

1. The docs in graphql-crunch are targeting `apollo-server-express`, not `express-graphql`. The latter doesn't have `formatResponse` ([though there is a PR](https://github.com/graphql/express-graphql/pull/404)). That means this implementation doesn't even work as is. 
2. This isn't written in a way that's testable in isolation. I decided to go down the route of creating a custom middleware just for crunching and pulling that out into lib. Turns out `express-graphql` never calls `next()` as apart of its middleware... therefore it won't yield execution to another middleware in the chain. The issue was [reported here](https://github.com/graphql/express-graphql/issues/420) with no official word on fixes, but there are a [couple](https://github.com/graphql/express-graphql/issues/279) [workarounds](https://github.com/graphql/express-graphql/issues/374).

The ultimate solution I came up with isn't quite as hacky feeling as the things mentioned in the end of 2, but I'd definitely value input on the approach.

I found a library called [express-interceptor](https://www.npmjs.com/package/express-interceptor) which gives you the ability to intercept and alter the body of a response before it's returned from express. This allows us to add the middleware _before_ `express-graphql` but have the crunching happen after `express-graphql` has resolved. I do have a concern about having to parse and re-stringify the body in the interceptor, but I'm not sure how big of an issue that will be. 

As for testing, I added a library called [supertest](https://www.npmjs.com/package/supertest) that allows the assertion of a request to express or similar http frameworks. It's an integration test of sorts because it tests the flow from `express` to `express-graphql` then to `graphql-crunch` via the interceptor. The good news is that it's a test we can run in jest so it's not going to require a real live environment to test against.

zephraph commented on src/lib/__tests__/crunchInterceptor.test.js:
```diff
@@ -0,0 +1,60 @@
+      .get("/?query={__schema{types{name}}}&crunch")
+      .set("Accept", "application/json")
+      .expect(res => {
+        expect(Array.isArray(res.body.data)).toBeFalsy()
```
The expectation is that a crunched result will be an array whereas a non-crunched result will be an object. This isn't a very sophisticated test, but it at least provides some level of validation. I'm not convinced this should even be a thing though... more on that below.

orta commented on src/lib/crunchInterceptor.js:
```diff
@@ -0,0 +1,13 @@
+    }
+    send(JSON.stringify(body))
+  },
+}))
```
I'm pretty sure the `__schema` comes from the official recommendation on generating a full introspection query, basically to generate the data that fills GraphiQL.

orta replied:
Ah yeah, good point, those clients wouldn't be adding the crunch param, so wouldn't see it at all

zephraph replied:
Even if a client did do an introspection query with the crunch flag, so long as the uncrunch step happened on the client before it was based to the graphql layer I'd think that it wouldn't matter. I reached out to @jamesreggio, maybe I'm just missing something.

jamesreggio replied:
Hey there — just confirmed that including the `__schema` condition was an unnecessary mistake.

We built `graphql-crunch` as a drop-in replacement for [`graphql-deduplicator`](https://github.com/gajus/graphql-deduplicator), which is a [lossy](https://github.com/prismagraphql/graphql-yoga/issues/303#issuecomment-386383356) compressor that depends upon Apollo-specific behavior. The `graphql-deduplicator` example code includes that condition, and we accidentally cargo-culted it into our README.

All that to say, it's safe to remove.

zephraph replied:
Sweet, thanks!

orta replied:
Makes sense 👋

zephraph commented:
Okay, I think this is ready for final review. 

I cleaned up the `crunchInterceptor` test file and created helpers in `src/tests` to assist in setting up a `graphql-express` instance and mocking out interceptors. I also added another test to ensure the interceptor isn't called on an error case. Added response checks on each of the existing tests. 

Let me know what you think.

zephraph commented:
I think there's a lot we could do around testing in general. It's something I'd love to talk more about after I start. Until then, this has been a fun intro to metaphysics. :smile:

mzikherman commented:
ok, merging! 😎 Great work!

zephraph commented:
I'll open a PR.
